### PR TITLE
Test io.ascii and table.Table write-read round-tripping

### PR DESCRIPTION
**Pull request by wmwv**
Adds a new test `test_write_read.py` that tests that all `test_defs` samples in `test_write.py` can be successfully written (to a StringIO object) and read back in to recover the correct data.

Note that adding this test creates 3 new failures.  This is a "success" of this feature branch.  But will thus break master passing its tests.

The failures are 
- #5160 - "Table fails to round-trip `ascii.aastex` due to incorrect requirement for a '\' on the last line of a table."  There are two examples in the suite that trigger this.
- #5205 - "LaTeX reader fails to read LaTeX table with astropy.units that was written with LaTeX writer"
- An error reading in the `\begin{tabletype}[tablealign]` example.  I'm a bit more stuck on what will need to be fixed for this.  I've never seen this LaTeX table type.

The first two are the motivating reasons why I wrote this round-tripping test in the first place, so it's great that only other one thing came up.
- Should I xfail these for now?

[*] The log shows 12 failures, but these are each of the key 3 failures failing in each of `fast_writer=True, False` and in each of the `io.ascii` and `table.Table` interfaces.

### CODE REVIEW

**Comment by wmwv** [comment]
@hamogu Here's the test round-trip write-read function.  I would very much welcome your comments and feedback.

**Comment by wmwv** [comment]
@taldcroft I wrote a round-tripping test function for `io.ascii` to test that `astropy.io.ascii` (and the `astropy.table.Table` interface to `io.ascii`) can read what it writes.  @hamogu kindly offered to take a look at it.  

I appreciate your support of this concept and would welcome your feedback as well.

File: astropy/io/ascii/tests/test_write.py [hunk 1]
```diff
@@ -339,7 +340,7 @@
          ),
 ]
 
-test_def_masked_fill_value = [
+test_defs_masked_fill_value = [
```
Comment (hamogu) [inline astropy/io/ascii/tests/test_write.py]: Why this change? It sees to me that "test_def" is jsut as good as "test_defs" here.

File: astropy/io/ascii/tests/test_write.py [hunk 2]
```diff
@@ -15,21 +15,22 @@
 from .common import setup_function, teardown_function
 
 test_defs = [
-    dict(kwargs=dict(),
+    dict(kwargs=dict(Writer=ascii.Basic),
```
Comment (hamogu) [inline astropy/io/ascii/tests/test_write.py]: Why do you specify the writer here? The intentions of this test is not only to test the basic writer, but also to make sure that the default option (which currently is the Basic writer) always gives the same answer. Most users actually don't set the Writer by hand, they just rely on the default, so we should test it that way, too.

File: astropy/io/ascii/tests/test_write_read.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        print('Expected:\n%s' % test_def['out'])
+        print('Actual:\n%s' % in_out.getvalue())
+        print('Read:\n%s' % str(foo))
+
+    assert [x.strip() for x in in_out.getvalue().strip().splitlines()] == \
+           [x.strip() for x in test_def['out'].strip().splitlines()]
+
+
+def check_read_write_table_via_table(test_def, table, fast_writer):
+    """Check reading and writing using the astropy.table.Table interface."""
```
Comment (hamogu) [inline astropy/io/ascii/tests/test_write_read.py]: I don't think both interfaces need to be tested separately. There are other tests to ensure that they all work properly. In principle, more tests are always better, but the astropy tests get run a lot on Travis and already take quite a while, so we should limit ourselves to tests that are strictly necessary.

File: astropy/io/ascii/tests/test_write_read.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    return False
+
+
+check_functions = \
+    (check_read_write_table_via_ascii, check_read_write_table_via_table)
+
+
+@pytest.mark.parametrize("test_def", test_defs)
+@pytest.mark.parametrize("check_function", check_functions)
+@pytest.mark.parametrize("fast_writer", [True, False])
```
Comment (hamogu) [inline astropy/io/ascii/tests/test_write_read.py]: definitely a good use of parameterize here to get a lot of cases covered with very little new code.

**Comment by hamogu** [comment]
I think this is a good think to test, specifically as we add features (e.g. the LaTeX writer used to round-trip in the past before we added support for units).

Can we add fixes for the test failures to this PR though? It's not going to get merged with the tests passing. You can say "But it's already broken, the tests only confirm that." The problem is that then all tests for all other PRs would also show a "failed" status which makes it a lot harder to spot the really new failures.
Alternatively, you could mark these tests as "expected to fail", and activate them one by one in other PRs that fix round-tripping, but it would be a lot better to fix the problem when it's found. And, at least for the LaTeX tables you have the fix already in another PR.

**Comment by hamogu** [comment]
Lastly, can you squash the commits into one or two? (Let me know if you don't know how. I always for get, but I know where to look it up.)

File: astropy/io/ascii/tests/test_write.py (hunk 1 above)
Comment (wmwv) [inline astropy/io/ascii/tests/test_write.py]: For consistency with `test_defs` and `test_defs_no_data`.  The prefix `test_defs` denotes a collection, while the individual test definition is the singular `test_def`.

"Pluralize test_defs_masked_fill_value for consistency"

File: astropy/io/ascii/tests/test_write_read.py (hunk 3 above)
Comment (wmwv) [inline astropy/io/ascii/tests/test_write_read.py]: I was just copying the testing structure as it is in `test_write.py`.

A separate ticket to streamline table testing might be appropriate.  I'm not a particularly big fan of this "almost-the-same-thing" functions here and would certainly be happy to make this cleaner, but I don't fully understand what the expected inconsistencies might be.

File: astropy/io/ascii/tests/test_write.py (hunk 2 above)
Comment (wmwv) [inline astropy/io/ascii/tests/test_write.py]: Good question and catch!

The answer is actually because the reading of these tables fails without a format specification being passed to the read.  Thanks for prompting me to remember that.  Once I write that down it clearly supports exactly your point.

So I will remove these two added `Writers=ascii.Basic` entries in the `kwargs` dict.  (the one above and the next one).  But this means we now have two more test cases that fail to round-trip.
